File: haskell/4.hs
import Data.List (sort)

isPalindrome :: (Eq a) => [a] -> Bool
isPalindrome x = x == reverse x

nums :: [Int]
nums = reverse $ sort [x*y | x <- [999,998..0], y <- [999,998..0]]

main = (print . head . filter (isPalindrome . show)) nums

File: python/4.py
def isPalindrome(s):
    return s == ''.join(list(reversed(s)))

nums = sorted([x*y for x in range(1000) for y in range(x+1)], reverse=True)

for x in nums:
    if isPalindrome(str(x)):
        print(x)
        break
